File: internal/ledger/day.go
package ledger

import (
	"fmt"
	"sort"
	"strings"
	"time"
)

// Day represents all entries for a single day
type Day struct {
	Date       time.Time
	Entries    []*Entry
	ScreenTime string
	Journal    string // Markdown journal entry for the day
}

// NewDay creates a new Day instance
func NewDay(date time.Time) *Day {
	return &Day{
		Date:       date,
		Entries:    make([]*Entry, 0),
		ScreenTime: "",
		Journal:    "",
	}
}

// HasJournal returns true if the day has a journal entry
func (d *Day) HasJournal() bool {
	return d.Journal != ""
}

// AddEntry adds an entry to the day
func (d *Day) AddEntry(entry *Entry) {
	entry.ScreenTime = d.ScreenTime
	d.Entries = append(d.Entries, entry)
}

// RemoveEntry removes an entry by ID
func (d *Day) RemoveEntry(id string) *Entry {
	for i, e := range d.Entries {
		if e.ID == id {
			removed := e
			d.Entries = append(d.Entries[:i], d.Entries[i+1:]...)
			return removed
		}
	}
	return nil
}

// GetEntry returns an entry by ID
func (d *Day) GetEntry(id string) *Entry {
	for _, e := range d.Entries {
		if e.ID == id {
			return e
		}
	}
	return nil
}

// UpdateEntry updates an existing entry
func (d *Day) UpdateEntry(entry *Entry) bool {
	for i, e := range d.Entries {
		if e.ID == entry.ID {
			d.Entries[i] = entry
			return true
		}
	}
	return false
}

// SetScreenTime sets the screen time for all entries in the day
func (d *Day) SetScreenTime(screenTime string) {
	d.ScreenTime = screenTime
	for _, e := range d.Entries {
		e.ScreenTime = screenTime
	}
}

// TotalCAD returns the sum of all CAD amounts
func (d *Day) TotalCAD() float64 {
	var total float64
	for _, e := range d.Entries {
		total += e.CAD
	}
	return total
}

// TotalIDR returns the sum of all IDR amounts
func (d *Day) TotalIDR() float64 {
	var total float64
	for _, e := range d.Entries {
		total += e.IDR
	}
	return total
}

// DateString returns the date formatted as YYYY-MM-DD (for CSV)
func (d *Day) DateString() string {
	return d.Date.Format("2006-01-02")
}

// DateDisplay returns the date formatted as MM/DD/YYYY
func (d *Day) DateDisplay() string {
	return d.Date.Format("01/02/2006")
}

// FormatDateDisplay returns a human-readable date format
func (d *Day) FormatDateDisplay() string {
	return d.Date.Format("January 2, 2006")
}

// EntryMatchesQuery checks if an entry matches the search query (vim-style)
// Searches across all fields: date, description, CAD, IDR
func EntryMatchesQuery(entry *Entry, query string) bool {
	if query == "" {
		return true
	}

	query = strings.ToLower(query)

	// Search in description
	if strings.Contains(strings.ToLower(entry.Description), query) {
		return true
	}

	// Search in date (multiple formats)
	dateStr := entry.Date.Format("01/02/2006")
	if strings.Contains(dateStr, query) {
		return true
	}
	dateStr2 := entry.Date.Format("1/2/2006")
	if strings.Contains(dateStr2, query) {
		return true
	}
	dateStr3 := strings.ToLower(entry.Date.Format("January 2, 2006"))
	if strings.Contains(dateStr3, query) {
		return true
	}
	dateStr4 := strings.ToLower(entry.Date.Format("Jan 2"))
	if strings.Contains(dateStr4, query) {
		return true
	}

	// Search in CAD amount
	cadStr := fmt.Sprintf("%.2f", entry.CAD)
	if strings.Contains(cadStr, query) {
		return true
	}

	// Search in IDR amount
	idrStr := fmt.Sprintf("%.0f", entry.IDR)
	if strings.Contains(idrStr, query) {
		return true
	}

	return false
}

// Filter returns entries matching the search query (vim-style, all fields)
func (d *Day) Filter(query string) []*Entry {
	if query == "" {
		return d.Entries
	}

	var filtered []*Entry
	for _, e := range d.Entries {
		if EntryMatchesQuery(e, query) {
			filtered = append(filtered, e)
		}
	}
	return filtered
}

// FilteredTotalCAD returns the sum of CAD for filtered entries
func (d *Day) FilteredTotalCAD(query string) float64 {
	var total float64
	for _, e := range d.Filter(query) {
		total += e.CAD
	}
	return total
}

// FilteredTotalIDR returns the sum of IDR for filtered entries
func (d *Day) FilteredTotalIDR(query string) float64 {
	var total float64
	for _, e := range d.Filter(query) {
		total += e.IDR
	}
	return total
}

// IsEmpty returns true if the day has no entries and no journal
func (d *Day) IsEmpty() bool {
	return len(d.Entries) == 0 && d.Journal == ""
}

// DateRange represents a range of days
type DateRange struct {
	Start time.Time
	End   time.Time
	Days  []*Day
}

// NewDateRange creates a new DateRange
func NewDateRange(start, end time.Time) *DateRange {
	return &DateRange{
		Start: start,
		End:   end,
		Days:  make([]*Day, 0),
	}
}

// AddDay adds a day to the range
func (dr *DateRange) AddDay(day *Day) {
	dr.Days = append(dr.Days, day)
	sort.Slice(dr.Days, func(i, j int) bool {
		return dr.Days[i].Date.Before(dr.Days[j].Date)
	})
}

// TotalCAD returns the sum of CAD for all days in the range
func (dr *DateRange) TotalCAD() float64 {
	var total float64
	for _, day := range dr.Days {
		total += day.TotalCAD()
	}
	return total
}

// TotalIDR returns the sum of IDR for all days in the range
func (dr *DateRange) TotalIDR() float64 {
	var total float64
	for _, day := range dr.Days {
		total += day.TotalIDR()
	}
	return total
}

// AllEntries returns all entries from all days, optionally filtered
func (dr *DateRange) AllEntries(query string) []*Entry {
	var entries []*Entry
	for _, day := range dr.Days {
		entries = append(entries, day.Filter(query)...)
	}
	return entries
}

// FilteredTotalCAD returns the sum of CAD for filtered entries across all days
func (dr *DateRange) FilteredTotalCAD(query string) float64 {
	var total float64
	for _, day := range dr.Days {
		total += day.FilteredTotalCAD(query)
	}
	return total
}

// FilteredTotalIDR returns the sum of IDR for filtered entries across all days
func (dr *DateRange) FilteredTotalIDR(query string) float64 {
	var total float64
	for _, day := range dr.Days {
		total += day.FilteredTotalIDR(query)
	}
	return total
}

// FormatRangeDisplay returns a human-readable range format
func (dr *DateRange) FormatRangeDisplay() string {
	return dr.Start.Format("01/02/2006") + " - " + dr.End.Format("01/02/2006")
}

// DayMatchesQuery checks if a day matches the query (for filtering entire days)
func DayMatchesQuery(day *Day, query string) bool {
	if query == "" {
		return true
	}

	query = strings.ToLower(query)

	// Check date
	dateStr := day.Date.Format("01/02/2006")
	if strings.Contains(dateStr, query) {
		return true
	}
	dateStr2 := day.Date.Format("1/2/2006")
	if strings.Contains(dateStr2, query) {
		return true
	}
	dateStr3 := strings.ToLower(day.Date.Format("January 2, 2006"))
	if strings.Contains(dateStr3, query) {
		return true
	}

	// Check screen time
	if strings.Contains(strings.ToLower(day.ScreenTime), query) {
		return true
	}

	// Check if any entry matches
	for _, e := range day.Entries {
		if EntryMatchesQuery(e, query) {
			return true
		}
	}

	return false
}

File: src/types/day.ts
import type { Entry } from './entry';
import { cloneEntry } from './entry';

/**
 * Day represents all entries for a single day
 */
export interface Day {
  date: Date;
  entries: Entry[];
  screenTime: string;
  journal: string; // Markdown journal entry for the day
}

/**
 * Creates a new Day instance
 */
export function createDay(date: Date): Day {
  return {
    date,
    entries: [],
    screenTime: '',
    journal: '',
  };
}

/**
 * Checks if the day has a journal entry
 */
export function hasJournal(day: Day): boolean {
  return day.journal !== '';
}

/**
 * Adds an entry to the day
 */
export function addEntry(day: Day, entry: Entry): void {
  entry.screenTime = day.screenTime;
  day.entries.push(entry);
}

/**
 * Removes an entry by ID, returns the removed entry or undefined
 */
export function removeEntry(day: Day, id: string): Entry | undefined {
  const index = day.entries.findIndex(e => e.id === id);
  if (index === -1) return undefined;
  const [removed] = day.entries.splice(index, 1);
  return removed;
}

/**
 * Gets an entry by ID
 */
export function getEntry(day: Day, id: string): Entry | undefined {
  return day.entries.find(e => e.id === id);
}

/**
 * Updates an existing entry
 */
export function updateEntry(day: Day, entry: Entry): boolean {
  const index = day.entries.findIndex(e => e.id === entry.id);
  if (index === -1) return false;
  day.entries[index] = entry;
  return true;
}

/**
 * Sets the screen time for all entries in the day
 */
export function setScreenTime(day: Day, screenTime: string): void {
  day.screenTime = screenTime;
  for (const entry of day.entries) {
    entry.screenTime = screenTime;
  }
}

/**
 * Returns the sum of all CAD amounts
 */
export function totalCAD(day: Day): number {
  return day.entries.reduce((sum, e) => sum + e.cad, 0);
}

/**
 * Returns the sum of all IDR amounts
 */
export function totalIDR(day: Day): number {
  return day.entries.reduce((sum, e) => sum + e.idr, 0);
}

/**
 * Checks if an entry matches the search query (vim-style)
 */
export function entryMatchesQuery(entry: Entry, query: string): boolean {
  if (!query) return true;

  const lowerQuery = query.toLowerCase();

  // Search in description
  if (entry.description.toLowerCase().includes(lowerQuery)) {
    return true;
  }

  // Search in date (multiple formats)
  const dateFormats = [
    formatDate(entry.date, 'MM/DD/YYYY'),
    formatDate(entry.date, 'M/D/YYYY'),
    formatDate(entry.date, 'MMMM D, YYYY'),
    formatDate(entry.date, 'MMM D'),
  ];
  for (const dateStr of dateFormats) {
    if (dateStr.toLowerCase().includes(lowerQuery)) {
      return true;
    }
  }

  // Search in CAD amount
  if (entry.cad.toFixed(2).includes(lowerQuery)) {
    return true;
  }

  // Search in IDR amount
  if (Math.round(entry.idr).toString().includes(lowerQuery)) {
    return true;
  }

  return false;
}

/**
 * Filters entries matching the search query
 */
export function filterEntries(day: Day, query: string): Entry[] {
  if (!query) return day.entries;
  return day.entries.filter(e => entryMatchesQuery(e, query));
}

/**
 * Returns the sum of CAD for filtered entries
 */
export function filteredTotalCAD(day: Day, query: string): number {
  return filterEntries(day, query).reduce((sum, e) => sum + e.cad, 0);
}

/**
 * Returns the sum of IDR for filtered entries
 */
export function filteredTotalIDR(day: Day, query: string): number {
  return filterEntries(day, query).reduce((sum, e) => sum + e.idr, 0);
}

/**
 * Returns true if the day has no entries and no journal
 */
export function isEmpty(day: Day): boolean {
  return day.entries.length === 0 && day.journal === '';
}

/**
 * Creates a deep copy of a day
 */
export function cloneDay(day: Day): Day {
  return {
    date: new Date(day.date),
    entries: day.entries.map(cloneEntry),
    screenTime: day.screenTime,
    journal: day.journal,
  };
}

// Helper function for date formatting (will be replaced by utils/format.ts)
function formatDate(date: Date, format: string): string {
  const months = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December',
  ];
  const shortMonths = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

  const m = date.getMonth();
  const d = date.getDate();
  const y = date.getFullYear();

  switch (format) {
    case 'MM/DD/YYYY':
      return `${String(m + 1).padStart(2, '0')}/${String(d).padStart(2, '0')}/${y}`;
    case 'M/D/YYYY':
      return `${m + 1}/${d}/${y}`;
    case 'MMMM D, YYYY':
      return `${months[m]} ${d}, ${y}`;
    case 'MMM D':
      return `${shortMonths[m]} ${d}`;
    default:
      return date.toISOString();
  }
}
